src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_CallerInfoAttributes.cs

```diff
@@ -690,6 +690,393 @@ static void Log(int p, [CallerArgumentExpression(p)] [CallerMemberName] string a
+
+class Program
+{
+    public static void Main()
+    {
+        var derived = new Derived();
+        derived.Log_AddAttributeInOverride(5 + 4);
+        derived.Log_RemoveAttributeInOverride(5 + 5);
+
+        ((Base)derived).Log_AddAttributeInOverride(5 + 4);
+        ((Base)derived).Log_RemoveAttributeInOverride(5 + 5);
+    }
+}
+";
+
+            var compilation = CreateCompilation(source, targetFramework: TargetFramework.NetCoreApp, options: TestOptions.ReleaseExe, parseOptions: TestOptions.Regular10);
+            CompileAndVerify(compilation, expectedOutput: @"5 + 4
+default
+default
+5 + 5").VerifyDiagnostics();
+        }
+
+        [ConditionalFact(typeof(CoreClrOnly))]
+        public void TestCallerArgumentWithUserDefinedConversionFromString()
+        {
+            string source = @"
+using System;
+using System.Runtime.CompilerServices;
+
+class C
+{
+    public C(string s) => S = s;
+    public string S { get; }
+    public static implicit operator C(string s) => new C(s);
+}
+
+class Program
+{
+    public static void Main()
+    {
+        Log(default(int));
+    }
+    const string p = nameof(p);
+    static void Log(int p, [CallerArgumentExpression(p)] C arg = null)
+    {
+        Console.WriteLine(arg.S);
+    }
+}
+";
+
+            var compilation = CreateCompilation(source, targetFramework: TargetFramework.NetCoreApp, options: TestOptions.ReleaseExe, parseOptions: TestOptions.Regular10);
+            compilation.VerifyDiagnostics(
+                // (19,29): error CS8959: CallerArgumentExpressionAttribute cannot be applied because there are no standard conversions from type 'string' to type 'C'
+                //     static void Log(int p, [CallerArgumentExpression(p)] C arg = null)
+                Diagnostic(ErrorCode.ERR_NoConversionForCallerArgumentExpressionParam, "CallerArgumentExpression").WithArguments("string", "C").WithLocation(19, 29)
+                );
+        }
+
+        [ConditionalFact(typeof(CoreClrOnly))]
+        public void TestCallerArgumentWithExtensionGetEnumerator()
+        {
+            string source = @"
+using System;
+using System.Collections;
+using System.Runtime.CompilerServices;
+
+public static class Extensions
+{
+    public static IEnumerator GetEnumerator(this IEnumerator enumerator, [CallerArgumentExpression(""enumerator"")] string s = ""default"")
+    {
+        Console.WriteLine(s);
+        return enumerator;
+    }
+}
+
+class Program
+{
+    static void Main()
+    {
+        var x = new [] { """", """" }.GetEnumerator();
+
+        foreach (var y in x)
+        {
+        }
+    }
+}
+
+";
+
+            var compilation = CreateCompilation(source, targetFramework: TargetFramework.NetCoreApp, options: TestOptions.ReleaseExe, parseOptions: TestOptions.Regular10);
+            CompileAndVerify(compilation, expectedOutput: "x").VerifyDiagnostics();
+        }
+
+        [ConditionalFact(typeof(CoreClrOnly))]
+        public void TestCallerArgumentWithExtensionDeconstruct()
```

Comment:
@333fred The compiler doesn't take a deconstruct extension into account if it has extra optional parameters. So this is an error-scenario. The same is true for index and range (I didn't add tests for them for now).